### ISSUE TITLE
ObjectDisposedException assertion in FileSystemWatcher tests

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#1124]
I sometimes get the following assert dialog when I run the unit tests for FileSystemWatcher locally:

```
---------------------------
Assertion Failed: Cancel=Debug, OK=Continue
---------------------------
ObjectDisposedException from something other than SafeHandle?

    at FileSystemWatcher.Monitor(Byte[] buffer)  C:\Users\stoub\Source\Repos\corefx\src\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs(173)
    at FileSystemWatcher.CompletionStatusChanged(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer)  C:\Users\stoub\Source\Repos\corefx\src\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs(365)
    at _IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)  
---------------------------
OK   Cancel   
---------------------------
```

**Comment by chcosta** [comment]
@ericstj , you did some investigation of this in the past.  I think the hope was that updating the IsHandleInvalid property to check the "IsClosed" property of _directoryHandle, would mitigate this (I'm no longer seeing failures on my box), but it appears that the race condition still exists.  Do you have any additional insight into this?  I'll start investigating this in depth.

**Comment by ericstj** [comment]
That helps since SafeHandle doesn't set IsInvalid after closing the handle, but there is still a race.  @stephentoub, is the stack of the ODE coming from the safehandle?

**Comment by chcosta** [comment]
What's the value of this assert?  It happens in a very small set of circumstances when some native process has disposed the handle.  Would it be more appropriate to just recover when the scenario has been detected?

**Comment by stephentoub** [comment]
@ericstj, yes, from DangerousAddRef. I added some code to the assert to print out the stack trace:
```
---------------------------
Assertion Failed: Cancel=Debug, OK=Continue
---------------------------
ObjectDisposedException from something other than SafeHandle?   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Microsoft.Win32.UnsafeNativeMethods.ReadDirectoryChangesW(SafeFileHandle hDirectory, Byte* lpBuffer, Int32 nBufferLength, Int32 bWatchSubtree, Int32 dwNotifyFilter, Int32& lpBytesReturned, NativeOverlapped* lpOverlapped, IntPtr lpCompletionRoutine)
   at System.IO.FileSystemWatcher.Monitor(Byte[] buffer) in c:\Users\stoub\Source\Repos\corefx\src\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 159
    at FileSystemWatcher.Monitor(Byte[] buffer)  c:\Users\stoub\Source\Repos\corefx\src\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs(172)
    at FileSystemWatcher.CompletionStatusChanged(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer)  c:\Users\stoub\Source\Repos\corefx\src\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs(366)
    at _IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)  
---------------------------
OK   Cancel   
---------------------------
```

**Comment by ericstj** [comment]
Yep, so it would seem this code is racing with a call to close on the handle.  The check for IsHandleInvalid returns false, but then someone closes the handle before this code runs.  I'm not aware of a pattern that enables addref-then-check on a disposed handle, so we'd either need to lock or swallow the exception.  I'd say we just swallow since that is what the code has been doing.  Just remove the assert since it isn't giving us any new information.

**Comment by chcosta** [comment]
Resolved with Issue #1150
